### PR TITLE
API Libraries > Python Substrate Interface

### PR DESCRIPTION
**Pull request by 0xLucca**
This page aims to add the Python Substrate Interface page

### CODE REVIEW

File: .snippets/code/develop/toolkit/api-libraries/py-substrate-interface/send_tx.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
+#...
+
+# 1. Compose the transfer call
```
Comment (nhussein11) [inline .snippets/code/develop/toolkit/api-libraries/py-substrate-interface/send_tx.py:3]:
```suggestion
# Compose the transfer call
```

File: .snippets/code/develop/toolkit/api-libraries/py-substrate-interface/send_tx.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+call = substrate.compose_call(
+    call_module="Balances",  # The pallet name
+    call_function="transfer_keep_alive",  # The extrinsic function
+    call_params={
+        'dest': 'INSERT_ADDRESS',  # Recipient's address
+        'value': 'INSERT_VALUE'  # Amount in smallest unit (e.g., Planck for DOT)
+    }
+)
+
+# 2. Create a signed extrinsic
```
Comment (nhussein11) [inline .snippets/code/develop/toolkit/api-libraries/py-substrate-interface/send_tx.py:13]:
```suggestion
# Create a signed extrinsic
```

File: .snippets/code/develop/toolkit/api-libraries/py-substrate-interface/send_tx.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+extrinsic = substrate.create_signed_extrinsic(
+    call=call, keypair=keypair  # Your keypair for signing
+)
+
+# 3. Submit and wait for inclusion
```
Comment (nhussein11) [inline .snippets/code/develop/toolkit/api-libraries/py-substrate-interface/send_tx.py:18]:
```suggestion
# Submit and wait for inclusion
```

File: .snippets/code/develop/toolkit/api-libraries/py-substrate-interface/substrate_interface.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
+from substrateinterface import SubstrateInterface
+
+# Connect to a node using websocket
+substrate = SubstrateInterface(
+    # For local node: "ws://127.0.0.1:9944"
+    # For Polkadot: "wss://rpc.polkadot.io"
+    # For Kusama: "wss://kusama-rpc.polkadot.io"
```
Comment (nhussein11) [inline .snippets/code/develop/toolkit/api-libraries/py-substrate-interface/substrate_interface.py:7]: I'd move these lines into a note above the snippet

File: develop/toolkit/api-libraries/py-substrate-interface.md [hunk 5]
```diff
@@ -1,4 +1,72 @@
...
+- Access to the account's keypair
+
+Here's how to create and submit a balance transfer:
+
+```py
+--8<-- 'code/develop/toolkit/api-libraries/py-substrate-interface/send_tx.py'
+```
+
+!!!note
+    The keypair object is essential for signing transactions. See the [Keypair documentation](https://polkascan.github.io/py-substrate-interface/usage/keypair-creation-and-signing/){target=\_blank} for more details.
```
Comment (nhussein11) [inline develop/toolkit/api-libraries/py-substrate-interface.md:62]:
```suggestion
    The `keypair` object is essential for signing transactions. See the [Keypair](https://polkascan.github.io/py-substrate-interface/usage/keypair-creation-and-signing/){target=\_blank} documentation for more details.
```

File: develop/toolkit/api-libraries/py-substrate-interface.md [hunk 6]
```diff
@@ -1,4 +1,72 @@
...
+
+## Next Steps
+
+Now that you understand the basics, you can:
+
+- Explore more complex queries and transactions
+- Learn about batch transactions and utility functions
+- Discover how to work with custom pallets and types
+
+For comprehensive documentation and advanced features, visit the [py-substrate-interface documentation](https://polkascan.github.io/py-substrate-interface/){target=\_blank}.
```
Comment (nhussein11) [inline develop/toolkit/api-libraries/py-substrate-interface.md:72]:
```suggestion
For comprehensive reference materials and advanced features, visit the [py-substrate-interface](https://polkascan.github.io/py-substrate-interface/){target=\_blank} documentation.
```

File: develop/toolkit/api-libraries/py-substrate-interface.md [hunk 7]
```diff
@@ -1,4 +1,72 @@
...
+
+The first step is to establish a connection to a Polkadot SDK-based node. You can connect to either a local or remote node:
+
+```py
+--8<-- 'code/develop/toolkit/api-libraries/py-substrate-interface/substrate_interface.py'
+```
+
+### Reading Chain State
+
+You can query various on-chain storage items. Here's how to check an account's balance and other details:
```
Comment (nhussein11) [inline develop/toolkit/api-libraries/py-substrate-interface.md:42]: Please add a sentence explaining that each query requires the specification of the pallet name, the storage item, and the required parameters 

So then adding this specific snippet would have context

**Review by nhussein11 — APPROVED** [review]
LGTM!

**Comment by eshaben** [comment]
@themacexpert can you please give this a formatting review?

File: develop/toolkit/api-libraries/py-substrate-interface.md [hunk 8]
```diff
@@ -1,4 +1,81 @@
...
+Here's how to create and submit a balance transfer:
+
+```py
+--8<-- 'code/develop/toolkit/api-libraries/py-substrate-interface/send_tx.py'
+```
+
+!!!note
+    The `keypair` object is essential for signing transactions. See the [Keypair](https://polkascan.github.io/py-substrate-interface/usage/keypair-creation-and-signing/){target=\_blank} documentation for more details.
+
+## Next Steps
```
Comment (eshaben) [inline develop/toolkit/api-libraries/py-substrate-interface.md:73]: Where to go Next?

**Review by themacexpert — APPROVED** [review]
Looks great!

File: develop/toolkit/api-libraries/py-substrate-interface.md [hunk 9]
```diff
@@ -1,4 +1,81 @@
...
+
+This Python library specializes in interfacing with a Polkadot SDK-based node, querying storage, composing extrinsics, SCALE encoding/decoding, and providing additional convenience methods to deal with the features and metadata of the Substrate runtime.
+
+The [Python Substrate Interface](https://github.com/polkascan/py-substrate-interface){target=\_blank} is a powerful library that enables interaction with Polkadot SDK-based chains. It provides essential functionality for:
+
+- Querying on-chain storage
+- Composing and submitting extrinsics
+- SCALE encoding/decoding
+- Interacting with Substrate runtime metadata
+- Managing blockchain interactions through convenient utility methods
```
Comment (eshaben) [inline develop/toolkit/api-libraries/py-substrate-interface.md:18]: This intro feels a bit repetitive. For example, you mention in the first sentence querying storage, composing extrinsics, SCALE encoding/decoding, convenience methods and then all of these items are repeated in the list.

Maybe just delete the first paragraph? 🤷‍♀️

File: develop/toolkit/api-libraries/py-substrate-interface.md [hunk 10]
```diff
@@ -1,4 +1,79 @@
...
+Here's how to create and submit a balance transfer:
+
+```py
+--8<-- 'code/develop/toolkit/api-libraries/py-substrate-interface/send_tx.py'
+```
+
+!!!note
+    The `keypair` object is essential for signing transactions. See the [Keypair](https://polkascan.github.io/py-substrate-interface/usage/keypair-creation-and-signing/){target=\_blank} documentation for more details.
+
+## Where to go next?
```
Comment (eshaben) [inline develop/toolkit/api-libraries/py-substrate-interface.md:71]:
```suggestion
## Where to Go Next
```

**Review by eshaben — COMMENTED** [review]
One last thing ✨

**Review by eshaben — APPROVED** [review]
LGTM!

**Pull request merged**